server/src/main/java/org/elasticsearch/index/search/MatchQuery.java

```diff
@@ -338,14 +404,35 @@ protected Query zeroTermsQuery() {
             this.mapper = mapper;
         }
 
+        public void setSynonymQueryStyle(SynonymQueryStyle synQueryStyle) {
+            this.synonymQueryStyle = synQueryStyle;
+        }
+
         @Override
         protected Query newTermQuery(Term term) {
             return blendTermQuery(term, mapper);
         }
 
         @Override
         protected Query newSynonymQuery(Term[] terms) {
-            return blendTermsQuery(terms, mapper);
+                switch (this.synonymQueryStyle) {
+                    case BEST_TERMS:
+                        List<Query> currPosnClauses = new ArrayList<Query>(terms.length);
+                        for (Term term : terms) {
+                            currPosnClauses.add(newTermQuery(term));
+                        }
+                        return new DisjunctionMaxQuery(currPosnClauses, 0.0f);
+                    case MOST_TERMS:
+                        BooleanQuery.Builder builder = new BooleanQuery.Builder();
+                        for (Term term : terms) {
+                            builder.add(newTermQuery(term), BooleanClause.Occur.SHOULD);
+                        }
+                        return builder.build();
+                    case BLENDED_TERMS:
+                        return blendTermsQuery(terms, mapper);
+                    default:
+                        throw new IllegalStateException("unrecognized synonymQueryStyle passed when creating newSynonymQuery");
```

Comment:
For easier debugging it would be good to print the (unknown) value of `this.synonymQueryStyle`